fs_attachment_s3_migration/models/ir_attachment.py

```diff
@@ -0,0 +1,252 @@
+# Copyright 2025 Cetmix OU
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import logging
+
+from odoo import api, models
+from odoo.osv.expression import AND, OR
+
+_logger = logging.getLogger(__name__)
+
+
+class IrAttachment(models.Model):
+    _inherit = "ir.attachment"
+
+    @api.model
+    def _s3_migration_domain(self, storage_code):
+        """Build domain for attachments to migrate, excluding force-DB files.
+
+        Respects force_db_for_default_attachment_rules to keep assets and
+        small images in database for performance.
+        """
+        base = [
+            ("store_fname", "not like", f"{storage_code}://%"),
+            "|",
+            ("res_field", "=", False),
+            ("res_field", "!=", False),
+        ]
+
+        force_db_domain = self._s3_get_force_db_domain(storage_code)
+        if force_db_domain:
+            return AND([base, ["!"] + force_db_domain])
+        return base
+
+    @api.model
+    def _s3_get_force_db_domain(self, storage_code):
+        """Get domain for attachments that must stay in DB for target storage.
+
+        Returns an Odoo domain combining MIME-type prefixes and optional
+        file_size limits. Each rule is normalized with AND() before OR'ing
+        with the accumulated domain to ensure correct grouping.
+        """
+        fs_storage = self.env["fs.storage"]
+        force_db_rules = fs_storage.get_force_db_for_default_attachment_rules(
+            storage_code
+        )
+        if not force_db_rules:
+            return []
+
+        domain = None
+        for mimetype_key, size_limit in force_db_rules.items():
+            rule_domain = [("mimetype", "=like", f"{mimetype_key}%")]
+            if size_limit:
+                rule_domain = AND([rule_domain, [("file_size", "<=", size_limit)]])
+            domain = OR([domain, rule_domain]) if domain else rule_domain
+
+        return domain or []
+
+    # ----------------------------------------------
+    # Migration helpers
+    # ----------------------------------------------
+    def _s3_resolve_migration_bytes(self, attachment, target_storage_code):
+        """Return non-empty base64 bytes for an attachment or None.
+
+        Prefer a donor already migrated to the target storage, then a donor
+        stored in DB. A donor is any attachment with the same checksum whose
+        ``datas`` can be read non-empty at this moment.
+        """
+        checksum = attachment.checksum
+        if not checksum:
+            return None
+
+        rf_domain = ["|", ("res_field", "=", False), ("res_field", "!=", False)]
```

Comment:
Same question about domain here. If this is needed on purpose, then there should be an explicit comment explaining the reason behind it.